engine/source/gui/game/guiAchievementPopupCtrl.cpp

```diff
@@ -0,0 +1,145 @@
+{
+    mIconBitmapName = StringTable->insert("");
+    mIconTextureObject = NULL;
+    mBackgroundBitmapName = StringTable->insert("marble/client/ui/achievement/background");
+    mBackgroundTextureObject = NULL;
+    mTitle = StringTable->insert("");
+    mHeader = StringTable->insert("Achievement Unlocked");
+}
+
+//------------------------------------------------------------------------------
+
+void GuiAchievementPopupCtrl::initPersistFields()
+{
+    Parent::initPersistFields();
+
+    addField("icon", TypeFilename, Offset(mIconBitmapName, GuiAchievementPopupCtrl));
+    addField("background", TypeFilename, Offset(mBackgroundBitmapName, GuiAchievementPopupCtrl));
+    addField("title", TypeString, Offset(mTitle, GuiAchievementPopupCtrl));
+    addField("header", TypeString, Offset(mHeader, GuiAchievementPopupCtrl));
+}
+
+//------------------------------------------------------------------------------
+
+bool GuiAchievementPopupCtrl::onWake()
+{
+    if (!Parent::onWake())
+        return false;
+
+    setIcon(mIconBitmapName);
+    setBackground(mBackgroundBitmapName);
+    setHeader(mHeader);
+
+    mProfile->constructBitmapArray();
+
+    return true;
+}
+
+void GuiAchievementPopupCtrl::setIcon(const char *name)
+{
+    mIconBitmapName = StringTable->insert(name);
+    if (*mIconBitmapName)
+        mIconTextureObject.set(mIconBitmapName, &GFXDefaultGUIProfile);
+    else
+        mIconTextureObject = NULL;
+
+    setUpdate();
+}
+
+void GuiAchievementPopupCtrl::setBackground(const char *name)
+{
+    mBackgroundBitmapName = StringTable->insert(name);
+    if (*mBackgroundBitmapName)
+        mBackgroundTextureObject.set(mBackgroundBitmapName, &GFXDefaultGUIProfile);
+    else
+        mBackgroundTextureObject = NULL;
+
+    setUpdate();
+}
+
+void GuiAchievementPopupCtrl::setTitle(const char *title)
+{
+    mTitle = StringTable->insert(title);
+    setUpdate();
+}
+
+void GuiAchievementPopupCtrl::setHeader(const char *header)
+{
+    mHeader = StringTable->insert(header);
+    setUpdate();
+}
+
+//------------------------------------------------------------------------------
+
+void GuiAchievementPopupCtrl::onRender(Point2I offset, const RectI &updateRect)
+{
+    if (mBackgroundTextureObject)
+    {
+        GFX->clearBitmapModulation();
+
+        RectI rect(offset, mBounds.extent);
+        GFX->drawBitmapStretch(mBackgroundTextureObject, rect);
+    }
+
+    if (mProfile->mBorder || !mBackgroundTextureObject)
+    {
+        RectI rect(offset, mBounds.extent);
+        GFX->drawRect(rect, mProfile->mBorderColor);
+    }
+
+    if (mIconTextureObject)
+    {
+        GFX->clearBitmapModulation();
+
+        RectI rect(offset.x + ACHIEVEMENT_ICON_OFFSET_X, offset.y + ACHIEVEMENT_ICON_OFFSET_Y, 64, 64);
+        GFX->drawBitmapStretch(mIconTextureObject, rect);
+    }
+
+    // I can't quite find an easy way to change the font size within this method, so both the achievement unlocked and
```

Comment:
While I would like this in the PR, if you can't figure it out, then we can deal with it later and make an issue for it.

Reply:
I might just be dumb cuz I still couldn't figure it out last time but I'll look into it a bit more